File: day8/8-2.js
import {nanomemoize as memoize} from 'nano-memoize';
import fs from 'fs';

// load input as string
const taskInput = fs.readFileSync('input.txt', 'utf8');

const moves = taskInput.split('\n')[0];
const nodes = {};
    taskInput.split('\n').slice(2).forEach(nodeLine => {
  const foundItems = nodeLine.split('=');
  const nodeKey = foundItems[0].trim();
  const nodeValue = [...foundItems[1].match(/(\w+)/g)];
  nodes[nodeKey] = nodeValue;
});

function applyMovesSingle(startingNode) {
  let currentNode = startingNode
  let stepCount = 0
  for (const move of moves){
    stepCount++;
    if (move === 'R') {
      currentNode = nodes[currentNode][1]
    } else {
      currentNode = nodes[currentNode][0]
    };
  };
  return [stepCount,currentNode];
}

const applyMovesSingleMemoized = memoize(applyMovesSingle)

function applyMoves(startingNodes) {
  let currentNodes = startingNodes;
  let stepCount = 0;

  for (const currentNode of currentNodes) {
    let newResult;
    newResult = applyMovesSingleMemoized(currentNode);
    stepCount += newResult[0];
    currentNodes[currentNodes.indexOf(currentNode)] = newResult[1];
  };
  stepCount = stepCount / startingNodes.length
  return stepCount;
}



let totalSteps = 0;
let currentNodes = Object.keys(nodes)
  .filter((key) => key[2] === "A")
const oneCycleLength = moves.length;


while (!(currentNodes.every(node => node[2] == 'Z'))) {
  totalSteps += applyMoves(currentNodes);
  if (totalSteps % (10000 * oneCycleLength) === 0) {
    console.log(totalSteps);
  }
}

console.log("FOUND");
console.log(totalSteps);

File: day8/8-2.py
import sys
import re
import math

# load input as array of strings
with open(sys.argv[1]) as file:
    taskInput = file.read()

moves = taskInput.split("\n")[0]
nodes = dict(
    [
        tuple(
            [
                foundItems[0]
                if len(foundItems := re.findall(r"(\w+)", nodeItem)) == 1
                else tuple(foundItems)
                for nodeItem in nodeLine.split("=")
            ]
        )
        for nodeLine in taskInput.split("\n")[2::]
    ]
)

sys.setrecursionlimit(100000000) # LMFAO

def countSteps(moves,startingNode,steps):
    currentNode = startingNode
    for move in moves:
        if currentNode[2] != 'Z':
            steps+=1
            if move == 'R':
                currentNode = nodes.get(currentNode)[1]
            else:
                currentNode = nodes.get(currentNode)[0]
    if currentNode[2] != 'Z':
        return countSteps(moves,currentNode,steps)
    else:
        return steps


totalSteps = 0
currentNodes = list(filter(lambda x: x[2] == "A", nodes))



cycleLength = math.lcm(*[countSteps(moves,node,0) for node in currentNodes])
print(cycleLength)